File: src/gmeow_tools/gts/writer.py
# SPDX-FileCopyrightText: 2026 Blackcat Informatics® Inc. <paudley@blackcatinformatics.ca>
# SPDX-License-Identifier: Apache-2.0
"""A GTS writer: build frames, maintain the id/prev chain, emit a CBOR Sequence.

This is the encoder counterpart to :mod:`gmeow_tools.gts.reader`. It drives the
conformance vectors and is the seed of the future ``RDF 1.2 → GTS`` producer.
"""

from __future__ import annotations

import cbor2

from gmeow_tools.gts.codec import DEFAULT_CATALOG, Codec, encode_chain
from gmeow_tools.gts.model import Quad, Term, Triple
from gmeow_tools.gts.wire import (
    MAGIC,
    SELF_DESCRIBE_TAG,
    VERSION,
    canonical,
    content_id,
    header_id,
)


def term_to_wire(t: Term) -> dict[str, object]:
    """Serialise a :class:`Term` to its wire map (dropping absent fields)."""
    out: dict[str, object] = {"k": int(t.kind)}
    if t.value is not None:
        out["v"] = t.value
    if t.datatype is not None:
        out["dt"] = t.datatype
    if t.lang is not None:
        out["l"] = t.lang
    if t.reifier is not None:
        out["rf"] = t.reifier
    return out


class Writer:
    """Accumulates a GTS log as a CBOR Sequence.

    Args:
        profile: The header ``"prof"`` value (§13).
        catalog: The transform catalog (id → :class:`Codec`).
        meta: Optional header metadata.
        magic_tag: Prefix the Header with the CBOR self-describe tag (§3).
    """

    def __init__(
        self,
        profile: str = "generic",
        catalog: dict[int, Codec] | None = None,
        meta: dict[str, object] | None = None,
        *,
        magic_tag: bool = True,
    ) -> None:
        """Initialise the writer and emit the Header (the chain genesis)."""
        self.catalog = catalog or dict(DEFAULT_CATALOG)
        self._name_to_id = {c.name: i for i, c in self.catalog.items()}
        header: dict[str, object] = {
            "gts": MAGIC,
            "v": VERSION,
            "prof": profile,
            "cat": {i: {"name": c.name, "cls": c.cls} for i, c in self.catalog.items()},
        }
        if meta is not None:
            header["meta"] = meta
        header["id"] = header_id(header)
        self._prev: bytes = header["id"]  # type: ignore[assignment]
        first = cbor2.CBORTag(SELF_DESCRIBE_TAG, header) if magic_tag else header
        self._buf = bytearray(canonical(first))

    @property
    def head(self) -> bytes:
        """The id the next appended frame must reference as ``"prev"``."""
        return self._prev

    def _chain_ids(self, chain: list[str] | None) -> list[int]:
        """Resolve codec names to file-local catalog ids."""
        return [self._name_to_id[name] for name in (chain or [])]

    def add_frame(
        self,
        frame_type: str,
        *,
        payload: object | None = None,
        raw: bytes | None = None,
        transform: list[str] | None = None,
        pub: object | None = None,
        to: list[dict[str, object]] | None = None,
        sig: bytes | None = None,
    ) -> bytes:
        """Append one frame and return its ``"id"``.

        Exactly one of ``payload`` (structured CBOR) or ``raw`` (blob bytes) is the
        payload source. When ``transform`` is given, the payload is CBOR-encoded (if
        structured) and run through the codec chain so ``"d"`` is a byte string (§6.1).
        """
        frame: dict[str, object] = {"t": frame_type}
        if transform:
            frame["x"] = self._chain_ids(transform)
            source = raw if raw is not None else canonical(payload)
            frame["d"] = encode_chain(transform, source)
        elif raw is not None:
            frame["d"] = raw
        elif payload is not None:
            frame["d"] = payload
        if pub is not None:
            frame["pub"] = pub
        if to is not None:
            frame["to"] = to
        frame["prev"] = self._prev
        frame["id"] = content_id(frame)
        if sig is not None:
            frame["sig"] = sig
        self._buf += canonical(frame)
        self._prev = frame["id"]  # type: ignore[assignment]
        return self._prev

    # -- convenience builders -------------------------------------------------

    def add_terms(
        self, terms: list[Term], *, transform: list[str] | None = None
    ) -> bytes:
        """Append a ``terms`` frame."""
        return self.add_frame(
            "terms", payload=[term_to_wire(t) for t in terms], transform=transform
        )

    def add_quads(
        self, quads: list[Quad], *, transform: list[str] | None = None
    ) -> bytes:
        """Append a ``quads`` frame (drops a ``None`` graph slot)."""
        rows = [
            [q[0], q[1], q[2], *([q[3]] if q[3] is not None else [])] for q in quads
        ]
        return self.add_frame("quads", payload=rows, transform=transform)

    def add_reifies(self, bindings: dict[int, Triple]) -> bytes:
        """Append a ``reifies`` frame binding reifier-ids to triples."""
        payload = {rid: list(spo) for rid, spo in bindings.items()}
        return self.add_frame("reifies", payload=payload)

    def add_annot(self, rows: list[Triple]) -> bytes:
        """Append an ``annot`` frame (reifier, predicate, value rows)."""
        return self.add_frame("annot", payload=[list(r) for r in rows])

    def add_blob(
        self,
        data: bytes,
        *,
        mt: str | None = None,
        rep: str | None = None,
        transform: list[str] | None = None,
    ) -> bytes:
        """Append an inline ``blob`` frame; metadata goes in ``pub`` (§12)."""
        pub: dict[str, object] = {}
        if mt is not None:
            pub["mt"] = mt
        if rep is not None:
            pub["rep"] = rep
        return self.add_frame("blob", raw=data, transform=transform, pub=pub or None)

    def add_meta(self, meta: dict[str, object]) -> bytes:
        """Append a ``meta`` frame."""
        return self.add_frame("meta", payload=meta)

    def add_suppress(
        self, targets: list[dict[str, object]], *, reason: str | None = None
    ) -> bytes:
        """Append a ``suppress`` frame (§11)."""
        payload: dict[str, object] = {"targets": targets}
        if reason is not None:
            payload["reason"] = reason
        return self.add_frame("suppress", payload=payload)

    def to_bytes(self) -> bytes:
        """Return the complete GTS file."""
        return bytes(self._buf)

File: crates/gts/src/writer.rs
// SPDX-FileCopyrightText: 2026 Blackcat Informatics® Inc. <paudley@blackcatinformatics.ca>
// SPDX-License-Identifier: Apache-2.0

//! A minimal GTS writer: build frames, maintain the id/prev chain, emit a CBOR
//! Sequence.
//!
//! This is the encoder counterpart to [`crate::reader`]. It currently supports
//! the frame types needed for the `files` profile (§13.2) and the conformance
//! vectors added with it. Deterministic CBOR and BLAKE3 self-hashes are handled
//! by [`crate::wire`].

use std::collections::HashMap;

use ciborium::value::Value;

use crate::codec::Codec;
use crate::model::{Quad, Term, Triple3};
use crate::wire::{canonical, content_id, digest_str, header_id, SELF_DESCRIBE_TAG};

fn iv(n: i64) -> Value {
    Value::Integer(ciborium::value::Integer::from(n))
}

/// Serialise a [`Term`] to its wire map (dropping absent fields).
pub fn term_to_wire(t: &Term) -> Value {
    let mut entries: Vec<(Value, Value)> = vec![("k".into(), iv(t.kind as i64))];
    if let Some(v) = &t.value {
        entries.push(("v".into(), v.clone().into()));
    }
    if let Some(dt) = t.datatype {
        entries.push(("dt".into(), iv(dt as i64)));
    }
    if let Some(l) = &t.lang {
        entries.push(("l".into(), l.clone().into()));
    }
    if let Some(rf) = t.reifier {
        entries.push(("rf".into(), iv(rf as i64)));
    }
    Value::Map(entries)
}

/// Accumulate a GTS log as a CBOR Sequence.
pub struct Writer {
    name_to_id: HashMap<String, i64>,
    prev: Vec<u8>,
    buf: Vec<u8>,
}

impl Writer {
    /// Create a writer and emit the Header (the chain genesis).
    pub fn new(profile: &str) -> Self {
        let catalog: HashMap<i64, Codec> = [
            (0i64, Codec { name: "identity".to_string(), cls: "encode".to_string() }),
            (1, Codec { name: "gzip".to_string(), cls: "compress".to_string() }),
            (2, Codec { name: "zstd".to_string(), cls: "compress".to_string() }),
            (7, Codec { name: "cose-encrypt0".to_string(), cls: "encrypt".to_string() }),
        ]
        .into_iter()
        .collect();
        let name_to_id: HashMap<String, i64> = catalog
            .iter()
            .map(|(id, c)| (c.name.clone(), *id))
            .collect();

        let cat_entries: Vec<(Value, Value)> = catalog
            .iter()
            .map(|(id, c)| {
                let mut ce: Vec<(Value, Value)> = vec![
                    ("name".into(), c.name.clone().into()),
                    ("cls".into(), c.cls.clone().into()),
                ];
                ce.sort_by(|a, b| canonical(&a.0).cmp(&canonical(&b.0)));
                (iv(*id), Value::Map(ce))
            })
            .collect();

        let mut header: Vec<(Value, Value)> = vec![
            ("gts".into(), "GTS1".into()),
            ("v".into(), iv(1)),
            ("prof".into(), profile.into()),
            ("cat".into(), Value::Map(cat_entries)),
        ];
        header.sort_by(|a, b| canonical(&a.0).cmp(&canonical(&b.0)));
        let id = header_id(&header);
        header.push(("id".into(), Value::Bytes(id.clone())));
        header.sort_by(|a, b| canonical(&a.0).cmp(&canonical(&b.0)));

        let tagged = Value::Tag(SELF_DESCRIBE_TAG, Box::new(Value::Map(header)));
        let buf = canonical(&tagged);

        Self {
            name_to_id,
            prev: id,
            buf,
        }
    }

    /// The id the next appended frame must reference as `"prev"`.
    pub fn head(&self) -> &[u8] {
        &self.prev
    }

    fn chain_ids(&self, chain: &[String]) -> Vec<i64> {
        chain
            .iter()
            .map(|name| self.name_to_id[name])
            .collect()
    }

    /// Append one frame and return its `"id"`.
    pub fn add_frame(
        &mut self,
        frame_type: &str,
        payload: Option<Value>,
        raw: Option<Vec<u8>>,
        transform: Option<&[String]>,
        pub_meta: Option<Value>,
    ) -> Vec<u8> {
        assert!(
            payload.is_none() || raw.is_none(),
            "payload and raw are mutually exclusive"
        );
        let mut frame: Vec<(Value, Value)> = vec![("t".into(), frame_type.into())];

        let data = match (transform, &payload, &raw) {
            (Some(chain), _, _) if !chain.is_empty() => {
                let source = raw.clone().unwrap_or_else(|| canonical(&payload.clone().unwrap()));
                // For the files profile we only need identity; compression is
                // intentionally not implemented in this minimal writer.
                assert!(
                    chain.iter().all(|n| n == "identity"),
                    "non-identity transforms require the Python producer"
                );
                let x_ids: Vec<Value> = self.chain_ids(chain).into_iter().map(iv).collect();
                frame.push(("x".into(), Value::Array(x_ids)));
                Value::Bytes(source)
            }
            (None, _, Some(r)) => Value::Bytes(r.clone()),
            (None, Some(p), None) => p.clone(),
            _ => Value::Null,
        };
        frame.push(("d".into(), data));

        if let Some(meta) = pub_meta {
            frame.push(("pub".into(), meta));
        }
        frame.push(("prev".into(), Value::Bytes(self.prev.clone())));

        frame.sort_by(|a, b| canonical(&a.0).cmp(&canonical(&b.0)));
        let id = content_id(&frame);
        frame.push(("id".into(), Value::Bytes(id.clone())));
        frame.sort_by(|a, b| canonical(&a.0).cmp(&canonical(&b.0)));

        self.buf.extend_from_slice(&canonical(&Value::Map(frame)));
        self.prev = id.clone();
        id
    }

    /// Append a `terms` frame.
    pub fn add_terms(&mut self, terms: &[Term]) -> Vec<u8> {
        let payload = Value::Array(terms.iter().map(term_to_wire).collect());
        self.add_frame("terms", Some(payload), None, None, None)
    }

    /// Append a `quads` frame (graph slot dropped when `None`).
    pub fn add_quads(&mut self, quads: &[Quad]) -> Vec<u8> {
        let rows: Vec<Value> = quads
            .iter()
            .map(|&(s, p, o, g)| {
                let mut row = vec![
                    iv(s as i64),
                    iv(p as i64),
                    iv(o as i64),
                ];
                if let Some(gv) = g {
                    row.push(iv(gv as i64));
                }
                Value::Array(row)
            })
            .collect();
        self.add_frame("quads", Some(Value::Array(rows)), None, None, None)
    }

    /// Append a `reifies` frame binding reifier-ids to triples.
    pub fn add_reifies(&mut self, bindings: &[(usize, Triple3)]) -> Vec<u8> {
        let mut map: Vec<(Value, Value)> = Vec::new();
        for (rid, (s, p, o)) in bindings {
            map.push((
                iv(*rid as i64),
                Value::Array(vec![
                    iv(*s as i64),
                    iv(*p as i64),
                    iv(*o as i64),
                ]),
            ));
        }
        self.add_frame("reifies", Some(Value::Map(map)), None, None, None)
    }

    /// Append an `annot` frame.
    pub fn add_annot(&mut self, rows: &[Triple3]) -> Vec<u8> {
        let rows: Vec<Value> = rows
            .iter()
            .map(|&(s, p, o)| {
                Value::Array(vec![
                    iv(s as i64),
                    iv(p as i64),
                    iv(o as i64),
                ])
            })
            .collect();
        self.add_frame("annot", Some(Value::Array(rows)), None, None, None)
    }

    /// Append an inline `blob` frame.
    pub fn add_blob(&mut self, data: &[u8], mt: Option<&str>) -> Vec<u8> {
        let pub_meta = mt.map(|m| {
            Value::Map(vec![("mt".into(), m.into())])
        });
        self.add_frame("blob", None, Some(data.to_vec()), None, pub_meta)
    }

    /// Append a `meta` frame.
    pub fn add_meta(&mut self, meta: Value) -> Vec<u8> {
        self.add_frame("meta", Some(meta), None, None, None)
    }

    /// Append a `suppress` frame.
    pub fn add_suppress(&mut self, targets: Vec<Value>, reason: Option<&str>) -> Vec<u8> {
        let mut payload: Vec<(Value, Value)> = vec![("targets".into(), Value::Array(targets))];
        if let Some(r) = reason {
            payload.push(("reason".into(), r.into()));
        }
        self.add_frame("suppress", Some(Value::Map(payload)), None, None, None)
    }

    /// Return the complete GTS file bytes.
    pub fn to_bytes(&self) -> Vec<u8> {
        self.buf.clone()
    }
}

/// Pack bytes into a `blake3:<hex>` digest string.
pub fn digest_string(data: &[u8]) -> String {
    digest_str(data)
}
